### DISCUSSION TITLE
Route.resource with optional :id for update, destroy, show methods

### DISCUSSION THREAD

**Discussion by ad-on-is** [adonisjs/core#2636] [Ideas]
I prefer using `Route.resource('endpoint').only(['index', 'update'])` over defining specific post, delete, etc. endpoints.

Sometimes I just need a DELETE or PUT request without an `:id` parameter.

Is there some kind of configuration of the Route.resource() mechanism to make that parameter optional?

From an API point of view, would it be propriate to modify AdonisJS to allow that?

**Comment by thetutlage** [comment]
How can you delete something without an id?

**Reply by bitkidd** [reply]
I guess he is referencing to https://guides.rubyonrails.org/routing.html#singular-resources

**Reply by ad-on-is** [reply]
Example 1:
If I want to allow users to delete their account, the resource would be `DELETE /users`.
This would be a UsersController `destroy` action which deletes multiple things, based on `auth.user.id`

Example 2:
If I want to add products to a wishlist.
`Route.resource('products.wishlist')` would give me a PUT `products/:product_id/wishlist/:id` endpoint, but the :id for the wishlist itself  in this case is unnecessary, since I want to update something based on the `:product_id` only.

**Reply by arthurcvm** [reply]
Hi @ad-on-is, in example 1 you can use middleware and validate if users have permission to do this.
In second example you need transform the Route.resource in a Route.group and personalize the PUT route sending all in request body.

**Reply by ad-on-is** [reply]
@arthurcvm My question is not about the technical implementation of the examples. I'm asking if their's anything wrong by making AdonisJS to handle the param :id optional for the generated put,patch,delete methods in Route.resource()

**Reply by thetutlage** [reply]
Resource follows a convention and if you don't like that convention, then you simply register the routes manually and don't resource at all.

Changing convention is not possible. 

However, as @bitkidd mentioned, there are singular resources which kind of works the way you want. But AdonisJS doesn't have a shortcut for them yet

**Reply by ad-on-is** [reply]
Ah, I see... thank you anyway 👍
